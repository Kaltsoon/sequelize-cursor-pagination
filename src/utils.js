let { Op } = require('sequelize');

if (!Op) {
  // Support older versions of sequelize
  Op = {
    and: '$and',
    or: '$or',
    lt: '$lt',
    lte: '$lte',
    gt: '$gt',
    gte: '$gte',
  };
}

const parseCursor = (cursor) => {
  if (!cursor) {
    return null;
  }

  try {
    return JSON.parse(Buffer.from(cursor, 'base64').toString('utf8'));
  } catch (e) {
    return null;
  }
};

const normalizePrimaryKeyField = (primaryKeyField) => {
  return Array.isArray(primaryKeyField) ? primaryKeyField : [primaryKeyField];
};

const ensurePrimaryKeyFieldInOrder = (order, primaryKeyField) => {
  const missingPrimaryKeyFields = primaryKeyField.filter(
    (pkField) => !order.find(([field]) => field === pkField),
  );

  return [...order, ...missingPrimaryKeyFields.map((field) => [field, 'ASC'])];
};

const normalizeOrder = (order, primaryKeyField, omitPrimaryKeyFromOrder) => {
  const normalizedPrimaryKeyField = normalizePrimaryKeyField(primaryKeyField);

  let normalized = [];

  if (Array.isArray(order)) {
    normalized = order.map((o) => {
      if (typeof o === 'string') {
        return [o, 'ASC'];
      }

      if (Array.isArray(o)) {
        const directionValueIndex = o.length - 1;
        o[directionValueIndex] = o[directionValueIndex] || 'ASC';

        return o;
      }

      return o;
    });
  }

  return omitPrimaryKeyFromOrder
    ? normalized
    : ensurePrimaryKeyFieldInOrder(normalized, normalizedPrimaryKeyField);
};

const reverseOrder = (order) => {
  return order.map((orderItem) => {
    orderItem[orderItem.length - 1] =
      orderItem[orderItem.length - 1].toLowerCase() === 'desc' ? 'ASC' : 'DESC';
    return orderItem;
  });
};

const serializeCursor = (payload) => {
  return Buffer.from(JSON.stringify(payload)).toString('base64');
};

const createCursor = (instance, order) => {
  // order
  // const fields
  const payload = order.map((orderItem) => {
    let field;
    if (typeof orderItem[0] == 'object') {
      return instance[orderItem[0]['as']][orderItem[1]];
    } else {
      field = orderItem[0];
    }

    return instance[field];
  });

  return serializeCursor(payload);
};

const isValidCursor = (cursor, order) => {
  return cursor.length === order.length;
};

const recursivelyGetPaginationQuery = (order, cursor) => {
  const directionValueIndex = order[0].length - 1;
  const currentOp =
    order[0][directionValueIndex].toLowerCase() === 'desc' ? Op.lt : Op.gt;

  // supporting only below format
  // [
  //   { model: Task, as: 'Task' },
  //   { model: Project, as: 'Project' },
  //   'createdAt',
  //   'DESC',
  // ];

  const _getColumnName = (order) => {
    // check if we have json object

    if (typeof order[0][0] === 'object') {
      const name = `${order[0][0]['as']}.${order[0][1]}`;
      return '$' + name + `$`;
    } else {
      return order[0][0];
    }
  };

  if (order.length === 1) {
    const key = _getColumnName(order);

    return {
      [key]: {
        [currentOp]: cursor[0],
      },
    };
  } else {
    const key = _getColumnName(order);

    return {
      [Op.or]: [
        {
          [key]: {
            [currentOp]: cursor[0],
          },
        },
        {
          [key]: cursor[0],
          ...recursivelyGetPaginationQuery(order.slice(1), cursor.slice(1)),
        },
      ],
    };
  }
};

const getPaginationQuery = (order, cursor) => {
  if (!isValidCursor(cursor, order)) {
    return null;
  }

  return recursivelyGetPaginationQuery(order, cursor);
};

module.exports = {
  Op,
  parseCursor,
  serializeCursor,
  normalizeOrder,
  isValidCursor,
  getPaginationQuery,
  createCursor,
  reverseOrder,
};
