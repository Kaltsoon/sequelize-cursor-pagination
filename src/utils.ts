import { Op, ModelStatic, WhereOptions, Model } from 'sequelize';
import { CursorPayload, OrderConfig } from './types';

export const parseCursor = (cursor: string): CursorPayload | null => {
  if (!cursor) {
    return null;
  }

  try {
    return JSON.parse(Buffer.from(cursor, 'base64').toString('utf8'));
  } catch (e) {
    return null;
  }
};

export const getPrimaryKeyFields = (model: ModelStatic<any>): string[] => {
  const primaryKeyFields = Object.entries(model.rawAttributes)
    .filter(([, attribute]) => attribute.primaryKey)
    .map(([column]) => column);

  return primaryKeyFields;
};

const normalizePrimaryKeyField = (
  primaryKeyField: string | string[],
): string[] => {
  return Array.isArray(primaryKeyField) ? primaryKeyField : [primaryKeyField];
};

const ensurePrimaryKeyFieldInOrder = (
  order: OrderConfig,
  primaryKeyField: string[],
): OrderConfig => {
  const missingPrimaryKeyFields = primaryKeyField.filter(
    (pkField) => !order.find(([field]) => field === pkField),
  );

  const primaryKeyOrder: OrderConfig = missingPrimaryKeyFields.map((field) => [
    field,
    'ASC',
  ]);

  return [...order, ...primaryKeyOrder];
};

export const normalizeOrder = (
  order: any,
  primaryKeyField: string | string[],
  omitPrimaryKeyFromOrder: boolean,
): OrderConfig => {
  const normalizedPrimaryKeyField = normalizePrimaryKeyField(primaryKeyField);

  let normalized = [];

  if (Array.isArray(order)) {
    normalized = order.map((o) => {
      if (typeof o === 'string') {
        return [o, 'ASC'];
      }

      if (Array.isArray(o)) {
        const [field, direction] = o;

        return [field, direction || 'ASC'];
      }

      return o;
    });
  }

  return omitPrimaryKeyFromOrder
    ? normalized
    : ensurePrimaryKeyFieldInOrder(normalized, normalizedPrimaryKeyField);
};

export const reverseOrder = (order: OrderConfig): OrderConfig => {
  return order.map(([field, direction]) => [
    field,
    direction.toLowerCase().split(' ')[0]  === 'desc' ? 'ASC' : 'DESC',
  ]);
};

const serializeCursor = (payload: CursorPayload): string => {
  return Buffer.from(JSON.stringify(payload)).toString('base64');
};

export const createCursor = <ModelType extends Model>(
  instance: ModelType,
  order: OrderConfig,
): string => {
  const payload = order.map(([field]) => instance.get(field));

  return serializeCursor(payload);
};

const isValidCursor = (cursor: CursorPayload, order: OrderConfig): boolean => {
  return cursor.length === order.length;
};

const recursivelyGetPaginationQuery = (
  order: OrderConfig,
  cursor: CursorPayload,
): WhereOptions<any> => {
  const currentOp = order[0][1].toLowerCase().split(' ')[0] === 'desc' ? Op.lt : Op.gt;

  if (order.length === 1) {
    return {
      [order[0][0]]: {
        [currentOp]: cursor[0],
      },
    };
  } else {
    return {
      [Op.or]: [
        {
          [order[0][0]]: {
            [currentOp]: cursor[0],
          },
        },
        {
          [order[0][0]]: cursor[0],
          ...recursivelyGetPaginationQuery(order.slice(1), cursor.slice(1)),
        },
      ],
    };
  }
};

export const getPaginationQuery = (
  order: OrderConfig,
  cursor: CursorPayload,
): WhereOptions<any> | null => {
  if (!isValidCursor(cursor, order)) {
    return null;
  }

  return recursivelyGetPaginationQuery(order, cursor);
};
