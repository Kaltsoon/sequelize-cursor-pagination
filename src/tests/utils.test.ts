import { Counter } from './models';
import { getPrimaryKeyFields, normalizeOrder } from '../utils';

describe('utils', () => {
  describe('normalizeOrder', () => {
    it('returns correct order with missing order', () => {
      expect(normalizeOrder(undefined, 'pk', false)).toMatchSnapshot();
    });

    it('returns correct order when primary key is omitted', () => {
      expect(
        normalizeOrder([['createdAt', 'DESC']], 'pk', true),
      ).toMatchSnapshot();
    });

    it('returns correct order with primary key in order', () => {
      expect(normalizeOrder([['pk', 'DESC']], 'pk', false)).toMatchSnapshot();
    });

    it('returns correct order with composite primary key', () => {
      expect(
        normalizeOrder([['a', 'DESC']], ['pk1', 'pk2'], false),
      ).toMatchSnapshot();
    });

    it('returns correct order with different orders', () => {
      expect(normalizeOrder(['a'], 'pk', false)).toMatchSnapshot();

      expect(
        normalizeOrder(
          [
            ['a', 'DESC'],
            ['b', 'ASC'],
          ],
          'pk',
          false,
        ),
      ).toMatchSnapshot();
    });

    it('returns correct order with different orders with nulls first', () => {
      expect(normalizeOrder(['a'], 'pk', false)).toMatchSnapshot();

      expect(
        normalizeOrder(
          [
            ['a', 'DESC NULLS FIRST'],
            ['b', 'ASC'],
          ],
          'pk',
          false,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('getPrimaryKeyFields', () => {
    test('returns correct primary key fields', () => {
      expect(getPrimaryKeyFields(Counter)).toEqual(['id']);
    });
  });
});
