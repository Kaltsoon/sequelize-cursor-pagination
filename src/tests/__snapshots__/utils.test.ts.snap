// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils normalizeOrder returns correct order when primary key is omitted 1`] = `
Array [
  Array [
    "createdAt",
    "DESC",
  ],
]
`;

exports[`utils normalizeOrder returns correct order with composite primary key 1`] = `
Array [
  Array [
    "a",
    "DESC",
  ],
  Array [
    "pk1",
    "ASC",
  ],
  Array [
    "pk2",
    "ASC",
  ],
]
`;

exports[`utils normalizeOrder returns correct order with different orders 1`] = `
Array [
  Array [
    "a",
    "ASC",
  ],
  Array [
    "pk",
    "ASC",
  ],
]
`;

exports[`utils normalizeOrder returns correct order with different orders 2`] = `
Array [
  Array [
    "a",
    "DESC",
  ],
  Array [
    "b",
    "ASC",
  ],
  Array [
    "pk",
    "ASC",
  ],
]
`;

exports[`utils normalizeOrder returns correct order with missing order 1`] = `
Array [
  Array [
    "pk",
    "ASC",
  ],
]
`;

exports[`utils normalizeOrder returns correct order with primary key in order 1`] = `
Array [
  Array [
    "pk",
    "DESC",
  ],
]
`;
